core/src/main/java/org/infinispan/container/offheap/BoundedOffHeapDataContainer.java

```diff
@@ -16,6 +16,7 @@
  * <p>
  * The link list consists of 28 bytes (3 longs and 1 int).  The first long is the actual entry address, the second is the
  * previous pointer, the third is the next pointer and lastly the int is the hashCode of the key to retrieve the lock.
+ * The hashCode is required    to know which lock to use when trying to read the entry.
```

Comment:
Updating docs.

Questions answered:
1. When an entry is removed or accessed and it is not the head or tail we have to update the before and after pointers. Without the previous node address, we would have to traverse the entire linked list to find the before.
2. This should be using something like malloc behind the scenes, but I don't know what various systems would allocate it as. At worse case I would hope the 28 bytes were allocated in a 32 byte spacing.
3. Do you mean instead of allocating memory for the entry and lru independently that we allocate 1 section of memory that holds both? I had originally done this. Unfortunately this cannot be done. The reason is that I can't read an entry unless I have at least the read lock stripe for that entry (or modify unless I have write lock). There is no reasonable way to store the hashCode for every single element in the lru list, since I would need them all to write lock before and after pointers to update all references properly. I hope this is enough to try to understand what I mean, but we can talk more if needed.
4. The size and `OffHeapDataContainer` is how many elements there are. The one in `BoundedOffHeapDataContainer` is either how many elements there are or how many bytes it is taking up. The latter is done in a way where it doesn't care which as it is handled by the `sizeCalculator` to determine which is used.

Reply:
1. Ok
2. I'd say 32 bytes is the minimum, malloc needs to store the allocation length inside the block and maybe some extra pointers (traditionally it uses linked lists to manage the free blocks, and the list pointer(s) should be in the blocks themselves).
3. Yeah, I'd like to talk more about this. I was thinking that the hash code should only change when the entry is released and replaced with something else, so it should be possible to use double-checked locking for the hash code. Access to the LRU pointers would still be guarded by `lruLock` instead of the entry locks.
4. I was thinking you could avoid the size update in OffHeapDataContainer for bounded containers, but doing it only for one `sizeCalculator` is probably too much.